### ISSUE TITLE
Rule and Command parameters

### ISSUE THREAD

**Issue by Aracon** [rolfvreijdenberger/izzum-statemachine#11]
Is it a good idea to add parameters to Rules and Commands?
I want to have Rules like "After X minutes" and Commands like "Send SMS" with parameters "minutes", "message text" in the transition configuration.
Maybe this is already possible in the current version?

**Comment by rolfvreijdenberger** [comment]
HI @Aracon ,
in essence it would be very nice to be able to add contextual parameters.
this would involve a statemachine configuration on a transition (or states) in a generic format (json?), which would be made available to be consumed by a rule or command.

I do see the use cases for this, but am reluctant to implement it. Is it the responsiblity of the statemachine to be able to configure the commands/rules? Or is that application logic?

The problem can be generalized in the rephrased version: "I'd like to be able to have a context in the rule or command that allows me to act conditionally on that context in the rule or command"

In case this context is provided to the rule/command, the application logic in that rule/command can take action on that context. THis could be achieved by injecting a Transition/State or even the Statemachine (though I do not like that at all) into the rule/command.


The main problem I have with this is that it makes the rule and command aware that it is being governed by the statemachine, making it tightly coupled (currently, the rules/commands are unaware of the statemachine) to a statemachine implementation

alternatively, a json string or the like could be injected when read from the statemachine transition/state configuration somewhere.
This could be implemented by having a seperate interface for the rule or command (IStatemachineAware or something along those lines).
When creating the rule/command for execution in the statemachine, we could check for that interface and inject the configuration/context in it.

Now, this COULD be implemented by a you:
- create a Loader subclass that can read the configuration for a rule (see https://github.com/rolfvreijdenberger/izzum-statemachine/blob/85f6976bfa1eb706fa3238eaff19417ffbd95d4b/src/statemachine/loader/Loader.php#L12)  and
- create a Transition subclass that holds the the configuration when loaded (see https://github.com/rolfvreijdenberger/izzum-statemachine/blob/85f6976bfa1eb706fa3238eaff19417ffbd95d4b/src/statemachine/Transition.php#L443, https://github.com/rolfvreijdenberger/izzum-statemachine/blob/85f6976bfa1eb706fa3238eaff19417ffbd95d4b/src/statemachine/Transition.php#L282)
- have these two play together to implement the desired behaviour by having the transition inject the loaded config at rule creation 
- depending on your backend of choice (php/json/yaml/sql/redis/mongo) this will be relatively easy to implement (php/yaml/json being the easiest I guess)

alternative courses of action:
- discuss options here, agree on a a generic implemenation accross the whole izzum package and implement
- subclass the statemachine and use the existing hooks to build it yourself

Let me know what you think or if the above is workable for you

cheeers Rolf

**Comment by Aracon** [comment]
It would be good to make a generic implementation. 
I think there is no need to inject a whole Transition/State, a json string (or PHP array) should be enough. I see this as an additional parameter in Transition configuration. This parameter is passed to the Rule constructor.
But I don't know how to handle the case of multiple rules in this approach. I mean the case when multiple Rules are chained in one transition configuration.
May be, when multiple rules are chained, then the parameter should be an array with same number of elements containing options for corresponding rules?

**Comment by rolfvreijdenberger** [comment]
a json string can be implemented in any backend so that would be my preferred choice.
the fact that it is possible to have chained rules and composite commands is a bad design choice on my part imo: you can achieve the same result in application logic with a little less flexibility but simpler code for the statemachine. 

Making this change would break the current existing implementation and would mean a major version bump. If that was the case, I would also make other changes at the same time, since this would be the opportune moment.
- no composite commands (solve/reuse in application command: means a less flexible configuration)
- no chained rules (solve/reuse in application command: means a less flexible configuration)
- falseRule ->IsFalse
- TrueRule ->IsTrue
- substates for the exit and entry commands (so you're able to retry failures on an exit or entry command)
- expose configuration - via an interface for both state entry and exit and transitions, for commands and rules - and put it on all backends. The specific rule or command that accepts the configuration can then delegate to existing rules/commands used within that rule command, without altering their interface. Or it could even be a specific StatemachineConfigCommand/Rule superclass to be extended, specifically to accept a configuration (default handling can be done here)
- probably simplify some of the approaches to use the statemachine
- maybe go full php7 on a branch ... (this would mean a seperate branch, where we would not have to be backward compatible) are you on php7?

There is some work involved here...
Would you like to join in the effort @Aracon ?

**Comment by rolfvreijdenberger** [comment]
the lack of having chained rules and composite commands will be offset by the fact that you can highly configure your commands/rules via a json configuration or the like.
This allows you to make runtime configuration changes without the need to do code adjustments

**Comment by Aracon** [comment]
Yes, I would like to work on implementation of command and rules config.
About other points in the changelist, I am not so sure. Some of them need time, I have not so much. My suggestion is that I will start working on command/rules config (and related changes) and then we will see how the work goes and decide about other changes.
I am still working on PHP 5.x

**Comment by Aracon** [comment]
The project I worked on is paused, so I pause the work on improvements. I will notify when the project will continue.
